### ISSUE TITLE
clear workflow  example

### ISSUE THREAD

**Issue by fariza** [format-message/format-message#29]
I am pretty sure I'm missing something basic here. But I don't understand how is supposed to work in a full development cycle.

From my understanding:
1. Add strings to source code `formatMessage('My Account Preferences')`
2. Extract the strings to a `json`, `yaml` to be passed to translators `format-message extract "src/**/*.js" > ./locales/en/messages.json`
3. Load those strings into your code depending on your locale by `formatMessage.setup(....`

Now the problem is, if add a new string, the step2 will overwrite all the strings, how do I just add some? The same for removing?
Basically the problem is how to handle the evolution of strings and its translations.

**Comment by thetalecrafter** [comment]
The idea is to always just add/remove strings directly in the source code, and then re-extract. You could run the extract as part of each commit that changes source strings, but it's more likely you would want to batch changes, and only send updated extracted messages to translators periodically, usually in preparation for a release.

As a product develops it is expected to have the translations lag a little behind the source language. During that time you can set the `missingTranslation` option to `"ignore"` or `"warning"`. When preparing a release that must have all the translations you can up the option to `"warning"` or `"error"`.

**Comment by fariza** [comment]
Yes but when you re-extract the json's where you put your translations get overwritten

**Comment by thetalecrafter** [comment]
Oh, put your translations in a different file. You should have one file per locale, and usually you'll only load the current locale translations. You can create an index.js file that combines all the locales for convenience in working with some of the apis that expect all the translations.

**Comment by fariza** [comment]
Ok, thanks, I was expecting a workflow style `l10ns` or `django`

**Comment by fariza** [comment]
I forgot to tell you, awesome libs!

**Comment by thetalecrafter** [comment]
Thanks! I'm glad you find it useful. Be sure to let me know if you find any issues or have suggestions.

**Comment by fariza** [comment]
Unfortunately, I don't think I will be able to use it, the automatic string "database" is fundamental for my taste!

**Comment by thetalecrafter** [comment]
Please describe the full desired workflow, and I can probably make the changes needed to accommodate.

**Comment by fariza** [comment]
Let me try to explain

The extraction of the strings go to files that are treated like a database (a collection of files, not really a database but for simplicity let's call it a db), this means, if one string is added in code, the corresponding entry is added in the "db", if one string is removed from the code the corresponding entry in the "db" is removed.

1. I add a string in the code, nothing happens because it takes the default value (or warning, error, etc..), 
2. Update the "db", now the "db" has the entry for the string but no translations yet
3. I add the translation for the corresponding string. 
4. Compile the "db", the code will see the available translations and make it available for the choosen locale.
5. It will be possible to choose runtime the locale.

If you look at how l10ns implements the workflow it is more or less like that, but your treatment of strings seems superior to me.

**Comment by thetalecrafter** [comment]
Thanks! I'll think about ways to do this and take a look at how l10ns does it.

**Comment by thetalecrafter** [comment]
How would the following be?

Source code:
```js
var fm = require('format-message')
fm.setup({
  generateId: require('format-message-generate-id/underscored_crc32'),
  translations: require('./strings')
})
// ...
fm('Hello, {location}', { location: 'Montreal' })
```

```bash
$ format-message extract --format db -l en -g underscored_crc32 -o ./strings.json
```

strings.json:
```json
{
  "en": {
    "hello_location_12ef23": { "message": "Hello, {location}" }
  },
  "fr": {
    "hello_location_12ef23": { "message": "Bonjour, {location}" }
  }
}
```

Running `format-message extract` would only modify the section matching the `-l` locale.

Or would you prefer translations to live right next to the original?

```js
formatMessage.setup({
  generateId: require('format-message-generate-id/underscored_crc32'),
  translations: formatMessage.db(require('./strings'))
})
```

strings.json
```json
{
  "hello_location_12ef23": {
    "message": "Hello, {location}",
    "description": "Greeting shown on the first screen.",
    "translations": {
      "fr": "Bonjour, {location}",
      "pt": "Olá, {location}"
    }
  }
}
```

**Comment by fariza** [comment]
I think I prefer the first option, it is not even necesary for the extracted strings to live in the same file, it can be separate files, actually if you are using translation services it might be better to have different files for each language.

The important part, is the reconciliation between strings in code and strings in the db, if add one to code it gets added to the db without modifying the others, if remove/change one from code remove  (or migrate as l10ns proposes) from db without touching the others.
Also, it should be possible to add locales at any time, for example I launch with en/fr but sometime after I add es, so it should be trivial to generate the db for es.

One question, in your proposed `setup` how do I choose and change the current locale at runtime?
What happen if the given locale doesn't have a translation for a given string?

**Comment by thetalecrafter** [comment]
I guess I misunderstood what you are asking for. What you describe sounds a lot like how things currently work.

Do you want to be able to edit the English (or whatever source language) strings by editing the extracted json file? I was expecting you to change the source.

In the current system you extract to a file containing exactly what is in the source. You send that file to the translation service, and then save the returned translations in a new file. As your source changes, you re-run `extract` to update the default strings file. You resend the file through translation. All translation services I have used (which admittedly is few) only need to retranslate new and changed strings. Then you get back the updated translations files.

There are two ways to set the locale. `formatMessage.setup({ locale: 'de' })` Changes the default locale for all messages. Passing the third argument `formatMessage(message, args, locale)` uses the locale for just this message.

If you don't have a translation for a message by default a console warning is printed, and the default (one specified in the source) is used. You can customize this behavior with `missingTranslation: 'ignore' | 'warning' | 'error'` and `missingReplacement: 'some string to use instead'`.

**Comment by fariza** [comment]
The problem is not with the extraction, is with the database, you rewrite the full db at each extraction and the update of existing fields has to be done by hand.

**Comment by thetalecrafter** [comment]
So you expect that if you were to change a message from "Yes" to "Maybe" in the extracted file, and then re-run extraction, your change would be preserved. The source code is no longer the authority, the json file is?

**Comment by thetalecrafter** [comment]
Current workflow:

1. Add `formatMessage()` calls in your source code.
2. `$ format-message extract` to a file containing all default messages.
3. Send default messages to translators.
4. Save translations in new files. Ensure `setup({ translations })` has all the needed translations.
5. Add/edit/remove `formatMessage()` calls in your source code.
6. Go to step 2.

It is not expected for you to manually edit the default messages file. The source code is authoritative. Messages and their descriptions are managed where they are used. When a new source file calls `formatMessage()` with a new message, the next extract will add it to the json file. When a default messages is changed in the source code, the next extract will change it in the json file. If a source file using format-message is removed, the next extract will remove the obsolete messages.

The whole idea is that you don't manage a database. You just write source code.

**Comment by fariza** [comment]
The problem is 2,3,4 this implies checking what was added and what was deleted to update the files with translations.
It should be the same file, the file that contains the translations and the file that is the output of extract

**Comment by thetalecrafter** [comment]
You would like a single file that contains both the extracted strings and their translations? And the extractor should not overwrite any translations in that file?

**Comment by fariza** [comment]
Did you try l10ns?

**Comment by thetalecrafter** [comment]
I haven't tried it, but I have read through their documentation. They have an opaque database and a cli command to add translations, as well as a web interface for translators.

The pain points imho are that the actual text lives separate from where it is used, so it is easy to get the parameters wrong and to change a string first you find the key in your source, then run a cli tool to edit the actual displayed text. You have to manually manage identifiers for your messages.

The workflow there requires you to add the English (or whatever your default is), as well as any other translations after running `l10ns update`. `l10ns update` only adds and removes keys from the database.

In format-message the default message is in your source code instead of an identifier. You edit it by editing your source code. You don't have to come up with or manually manage identifiers for your messages unless you want to disambiguate otherwise equivalent messages. If you change a message the auto-generated identifier will also be different so an outdated translation will not be shown.

`format-message extract` currently outputs a file that exactly represents all of the default messages in your source code, so you can give it to your translators.

If I'm understanding you correctly, you would like it to just do what `l10ns update` does, that is add and remove keys/identifiers in a file/database with all translations. If you change the message text in the file/database, `extract` won't mess with it, so whatever default message is in the source code will at runtime be replaced even in the default locale.

That would allow you to run `extract` targeting each of the translation files to automatically add new messages, and remove obsolete messages, which could be very helpful in maintaining translations.

Am I getting any closer to understanding?

**Comment by czert** [comment]
Hi,

yes, I think you get the difference now. As I understand it, `format-messages` treats the source code as authoritative for the default translation and doesn't help much with maintaining the translations. `l10ns` forces developers to come up with keys for the translated strings and their default value is in fact just one of the translations.

For me at least, the best approach is the way gettext does it: you put the default strings into source code (like you do with `format-messages`) and then you generate PO files for translations. In these files, the default strings are the keys and you fill in the translation for a given language. 

When this default string changes (you fix a typo or whatever), the original translations don't get lost: using string comparison and source code line numbers (I think), gettext can figure out that you did not in fact just delete a string and add another, but that you changed one. So when you generate new PO files, it uses the new default string with its old translation and marks the translation as _fuzzy_. That flags the translation for review by translators, so that they can check whether you just fixed a typo or made a change that needs to be reflected in the translations.

(see https://www.gnu.org/software/gettext/manual/html_node/Fuzzy-Entries.html)

The result is the best of both worlds: there's no need for explicit keys, but you still have a robust workflow for maintaining the translation "database".

**Comment by fariza** [comment]
I agree, `gettext` has the best workflow for me. That kind of behavior was that I was looking for.

**Comment by thetalecrafter** [comment]
Fuzzy entries sound really cool. I don't really know where I'd start on implementing that, but I would happily accept a PR for it. ;)

It sounds like the most pain, and easiest remedy, is around managing translations. The `extract` tool is really only for creating the file for the default language, and there isn't anything for managing the translation files.

Creating a tool that adds new entries and removes obsolete entries in translation files would be a helpful first step. Then that can be built out to meet the needs of multiple workflows.

Sound good?

**Comment by czert** [comment]
Sounds great to me :)

Once we have such a tool, I could take a shot at implementing fuzzy entries.

**Comment by sullivanpt** [comment]
Pardon me for putting myself into the middle of the conversation. First, I want to say big thanks for putting this project together.  I especially like the way the message extraction integrates cleanly with a large variety of inline message declarations: the automatic or manual message ID, the optional description text.  I also love the full eco-system of tools: cli, eslint, babel plugins.  Really awesome!

But the workflow needs a ton of improvement. Even the first extraction need manual reformatting just to consume it's own output, and secondary extractions require even more manual labor for incremental translations.

I really think you should team up with this project (https://github.com/guigrpa/mady/issues/14).  It is has a largely compatible API surface and already has excellent round trip tooling for incremental translations. Put the two projects together, add a little extra support for offline translations such as import and export xliff, and I think we can finally put gettext envy to bed. :-).

I'd love to help make this happen.

**Comment by adjenks** [comment]
Yeah it confused me as to what you do on the second round of translations. In particular, step 6 of this workflow:
1. Put text patterns into code.
2. Extract patterns into default translation file. e.g. `en.json`
3. Copy default to locale files and translate them to the locale. e.g. `de.json`
4. Update text patters in original code, adding and removing things.
5. Extract again, obtaining new default translation file. e.g. `new-en.json`
6. How do you gracefully merge the new default translation with existing files? (`new-de.json`?)

I would want a tool to do this:
- Overwrite keys/ids that have changed with an untranslated copy
- Delete keys from translations that are no longer present in the new template
- Add keys from the new template that don't exist in the translation files.

I would imagine that you would need to preserve the original template and compare them in a process like this:
```js
import equal from 'fast-deep-equal';
var latestTemplate = await import('./new-en.json'),
oldTemplate = await import('./en.json'),
oldDE = await import('./de.json'),
newDE = Object.keys(latestKeys).reduce((a, c) => {
    a[c] = (null == oldTemplate[c])
        ? latestTemplate[c]
        : (
            equal(latestTemplate[c], oldTemplate[c])
                ? oldDE[c]
                : latestTemplate[c]
        )
    return a
}, {})
//Overwrite de.json with newDE
```
So newDE would then have:
- The latest translation keys added
- The keys that changed between the old **template** and the new template would be reset to untranslated
- Old keys not not present in the new template would be discarded

Is this the process that you're hoping for @fariza ?

**Comment by vanwagonet** [comment]
I'm afraid I haven't prioritized any work on this since I switched to app development years ago. I'm open to review changes and give permissions to a new maintainer, but I won't be making changes myself, as I no longer use the library or spend much time in the JavaScript ecosystem.
